// tsparticles-interaction-external-connect/umd/Options/Classes/Connect.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./ConnectLinks"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.Connect = void 0;
    const ConnectLinks_1 = require("./ConnectLinks");
    class Connect {
        constructor() {
            this.distance = 80;
            this.links = new ConnectLinks_1.ConnectLinks();
            this.radius = 60;
        }
        get lineLinked() {
            return this.links;
        }
        set lineLinked(value) {
            this.links = value;
        }
        get line_linked() {
            return this.links;
        }
        set line_linked(value) {
            this.links = value;
        }
        load(data) {
            if (!data) {
                return;
            }
            if (data.distance !== undefined) {
                this.distance = data.distance;
            }
            this.links.load(data.links ?? data.lineLinked ?? data.line_linked);
            if (data.radius !== undefined) {
                this.radius = data.radius;
            }
        }
    }
    exports.Connect = Connect;
});
